oauth_session.go

```diff
@@ -0,0 +1,76 @@
+// Copyright 2012 Jimmy Zelinskie. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package reddit implements an abstraction for the reddit.com API.
+package geddit
+
+import (
+	"encoding/json"
+	"net/http"
+
+	"golang.org/x/net/context"
+	"golang.org/x/oauth2"
+)
+
+// OAuthSession represents an OAuth session with reddit.com --
+// all authenticated API calls are methods bound to this type.
+type OAuthSession struct {
+	Client       *http.Client
+	ClientID     string
+	ClientSecret string
+	Username     string
+	Password     string
+	OAuthConfig  *oauth2.Config
+}
+
+// NewLoginSession creates a new session for those who want to log into a
+// reddit account via OAuth.
+func NewOAuthSession(username, password, useragent, clientID, clientSecret string) (*OAuthSession, error) {
+	s := &OAuthSession{
+		Username: username,
+		Password: password,
+	}
+
+	s.OAuthConfig = &oauth2.Config{
+		ClientID:     clientID,
+		ClientSecret: clientSecret,
+		Scopes:       []string{}, //"identity", "read"},
+		Endpoint: oauth2.Endpoint{
+			AuthURL:  "https://oauth.reddit.com",
+			TokenURL: "https://www.reddit.com/api/v1/access_token",
+		},
+	}
+
+	ctx := context.Background()
+
+	// TODO offer auth code version as well as personal scripts
+	t, err := s.OAuthConfig.PasswordCredentialsToken(ctx, s.Username, s.Password)
+	if err != nil {
+		return nil, err
+	}
+
+	s.Client = s.OAuthConfig.Client(ctx, t)
+	return s, nil
+}
+
+func (s OAuthSession) Me() (*Redditor, error) {
```

Comment:
Probably breaking these endpoint functions down to bite-size pieces is best, considering there are so many API endpoints.